torchmultimodal/modules/layers/transformer_decoder.py

```diff
@@ -0,0 +1,199 @@
+        self.mlp_block = MLPBlock(embedding_dim, embedding_dim, [embedding_dim * 4])
+
+    def forward(self, x, decode_step, decode_idx):
+        h = self.pre_attention_norm(x)
+        if self.training:
+            h = checkpoint(self.attention, h, h, h, decode_step, decode_idx)
+        else:
+            h = self.attention(h, h, h, decode_step, decode_idx)
+        h = self.post_attention_dropout(h)
+        x = x + h
+
+        h = self.pre_mlp_norm(x)
+        if self.training:
+            h = checkpoint(self.mlp_block, h)
+        else:
+            h = self.mlp_block(h)
+        h = self.post_mlp_dropout(h)
+        x = x + h
+
+        return x
+
+
+class TransformerDecoder(nn.Module):
+    def __init__(
+        self,
+        position_embedding_input,
+        position_embedding_output,
+        num_attention_heads,
+        num_hidden_layers,
+        dropout,
+        attention_dropout,
+    ):
+        super().__init__()
+        if (
+            position_embedding_input.embedding_dim
+            != position_embedding_output.embedding_dim
+        ):
+            raise ValueError(
+                f"The position embedding dimension of the input modality {position_embedding_input.embedding_dim}"
+                f"must be equal to that of the output modality {position_embedding_output.embedding_dim}"
+            )
+
+        self.embedding_dim = position_embedding_input.embedding_dim
+        self.right_shift = RightShift(self.embedding_dim)
+        self.position_embedding_input = position_embedding_input
+        self.position_embedding_output = position_embedding_output
+        self.input_seq_len = self.position_embedding_input.seq_len
+        self.output_seq_len = self.position_embedding_output.seq_len
+        total_len = self.input_seq_len + self.output_seq_len
+
+        self.attention_nets = nn.ModuleList(
+            [
+                TransformerLayer(
+                    shape=(total_len),
+                    embedding_dim=self.embedding_dim,
+                    num_attention_heads=num_attention_heads,
+                    num_hidden_layers=num_hidden_layers,
+                    dropout=dropout,
+                    attention_dropout=attention_dropout,
+                )
+                for i in range(num_hidden_layers)
+            ]
+        )
+
+    def _get_inference_embeddings(self, x, position_embedding, decode_step):
+        # Use the data point at ``decode_step - 1`` to generate the next.
+        return position_embedding(x, decode_step)
+
+    def _get_training_embeddings(self, x, position_embedding):
+        # Use the full sequence for training
+        return x + position_embedding(x)
+
+    def forward(self, x_input, x_output, decode_step=None, decode_idx=None):
+        # Trigger inference mode if either of the input/output-modality sequence is ``None`` as generation
+        # is based the previous data point (autoregressive) where only one modality is needed at any point
```

Comment:
do we envision single/different decoder for autoregressive and NAR? good to think for unification cc: @RdoubleA